tests/test_config.py

```diff
@@ -51,3 +51,126 @@ def test_no_ini_path(self):
+
+        arguments = argparse.Namespace()
+        setattr(arguments, variable_name, arg_value)
+        value = config._get_config_value(
+            arguments,
+            variable_name,
+            settings={variable_name: "The worstest"},
+            env={env_name: "The worst stuff"},
+        )
+        self.assertEqual(arg_value, value)
+
+    def test_env_preferred(self):
+        """Test to see that env value is preferred when the variable exists in
+        settings as well"""
+        variable_name = "my-var"
+        env_value = "The best stuff around"
+        env_name = "CARAMEL_MY_VAR"
+        arguments = argparse.Namespace()
+        setattr(arguments, variable_name, None)
+        value = config._get_config_value(
+            arguments,
+            variable_name,
+            settings={variable_name: "The worstest"},
+            env={env_name: env_value},
+        )
+        self.assertEqual(env_value, value)
+
+    def test_settings_only(self):
+        """Test to see that the value from settings is returned when no argument or
+        env-variable"""
+        variable_name = "my-var"
+        settings_value = "Come on Toshi, you're sooo good"
+        arguments = argparse.Namespace()
+        setattr(arguments, variable_name, None)
+        value = config._get_config_value(
+            arguments,
+            variable_name,
+            settings={variable_name: settings_value},
+            env={},
+        )
+        self.assertEqual(settings_value, value)
+
+    def test_settings_only_no_argument(self):
+        """Test to see that the value from settings is returned when variable is not
+        settable from the commandline and no env-variable"""
+        variable_name = "my-var"
+        settings_value = "Come on Toshi, you're sooo good"
+        arguments = argparse.Namespace()
+        value = config._get_config_value(
+            arguments,
+            variable_name,
+            settings={variable_name: settings_value},
+            env={},
+        )
+        self.assertEqual(settings_value, value)
+
+    def test_different_setting_name_only(self):
+        """Test to see that the value from settings is returned when the name of the
+        settings differs from the varaible name"""
+        variable_name = "my-var"
+        setting_name = "something.very.different"
+        settings_value = "Come on Toshi, you're sooo good"
+        arguments = argparse.Namespace()
+        setattr(arguments, variable_name, None)
+        value = config._get_config_value(
+            arguments,
+            variable_name,
+            setting_name=setting_name,
+            settings={setting_name: settings_value},
+            env={},
+        )
+        self.assertEqual(settings_value, value)
+
+    def test_nothing(self):
+        """Test to see that if no value has been supplied as either an argument,
+        environment-variable or setting it returns None"""
+        variable_name = "my-var"
```

Comment:
I'd suggest reshuffling all the tests to do:

 
variable_name = ...
arguments = ...
setattr(arguments...)

simply to separate the boiler plate from the content. That makes it easier to read and compare the behaviour.

Reply:
this commit is from a bad rebase, pushed correct rebase